In [1]:
#Bitte ausführen, damit alles Notwendige importiert wird
#Note: Bei Änderungen der zugrundeliegenden Python-Files muss Jupyter neugestartet werden
import scipro

In [2]:
%%html
<!--Bitte diese Cell mit Run ausführen, damit die Styles geladen werden-->
<!--Bei Änderungen des CSS muss das Notebook im Browser neu geladen werden-->
<link rel="stylesheet" href="./styles/sciprolab.css">


# Scientific Programming Lab


## Vektoren

<div><img src="images/machine_learning_2x.png" width=300 align=right alt="xkcd comic describing machine learning as a pile of linear algebra, https://xkcd.com/1838/"/>
</div>    

### Ziele und Inhalte

- Mathmatische Inhalte
  - Definition von Vektoren, Vektoren im $\mathbb{R}^n$
  - Rechneoperationen mit Vektoren und Skalaren
  - Änhlichkeitsmaße 
- Informatische Inhalte
  - Arrays im allgemeinen
  - Arrays und ähnliches in Python
  - NumPy Arrays
  - Vektoren im Bereich *künstliche Intelligenz*

## Arrays 

*Arrays* oder *Felder* (der deutsche Begriff ist mehrdeutig, und meist importieren wir einfach das Fremdwort) gehören zu den ältesten Datenstrukturen, die es im Bereich der (praktischen) Informatik gibt. Ein 1-dimensionales Array ist eine Folge von im Speicher direkt nacheinander liegenden Datenelementen der selben Größe (und normalerweise des selben Typs). Der Zugriff auf Array-Elemente erfolgt normalerweise über den *Index* des gesuchten Elementes -  `arr[5]` würde so auf das Element mit Index 5 von `arr` zugreifen (normalerweise das *sechste* Element - siehe unten).

Man kann die Adresse jedes Elementes leicht aus der Anfangsadresse des Arrays und der Größe der Elemente errechnen - das ist eine *lineare Transformation*. Der Speicher eines normalen Rechners hat große Ähnlichkeiten mit einem Array - man kann ihn als Array von Bytes (oder von Worten) betrachten. Deswegen sind Arrays auf konventionellen Rechnern auch sehr leicht zu implementieren. Historisch gibt es verschiedene Indizierungsmethoden, aber heutzutage hat sich weitgehend die sehr einfache Methode durchgesetzt, für die sich auch schon die Entwickler der Sprache C entschieden haben: Arrays mit $n$ Einträgen haben Indizes im Bereich $0$ bis $(n-1)$. 

|arr |0|1|2|3|4|5|
|---|---|---|---|---|---|---|
| |4|8|15|16|23|42|

Es gibt eine Vielzahl von Anwendungen von Arrays - von Datenbanktabellen (mit vielen gleichartigen Zeilen) bis zu Klimamodellen (wo der Zustand der Atmosphäre in einem festen Raster simuliert wird). Gerade in letzterem Fall kommen auch mehrdimensionale Arrays zum Einsatz. Ein mehrdimensionales Array ist konzeptionell ein Array von Arrays geringerer Dimension. Insbesondere kann man 2-D-Arrays als recheckige Schemata von Datenelementen betrachten, wobei jede Zeile (bei "row-major Layout") ein 1-D-Array ist:

|arr |0|1|2|3|
|---|---|---|---|---|
|**0**|||||
|**1**||||17|
|**2**|||||
|**3**|||||
|**4**|||||

Der Zugriff auf z.B. das Element mit dem Wert 17 erfolgt dabei typischerweise mit `arr[1][3]` - in manchen Sprachen auch mit `arr[1,3]`. Da klassiche Arrays eine feste größe haben, kann man auch für mehrdimensionale Arrays die Adresse eines Datenelements in konstanter Zeit aus den Indices errechnen. Dieser schnelle Zugriff in **O**(1) ist eine wichtige Eigenschaft von *echten* Arrays.

Neben echten Arrays gibt es heute eine Vielzahl von anderen Datenstrukturen, die ein äußerlich ähnliches Interface anbieten - z.B. sogenannte *assoziative Arrays* und auch (speziell in Python) manche Implementierungen von *Listen*. Diese haben verschiedene Vor- und Nachteile, sind aber hier nicht unser Thema.

## Arrays und Array-ähnliche Datenstrukturen in Python

### Listen

Die häufigste sequentielle Datenstruktur in Python ist die *Liste*. Python-Listen sind sehr vielseitig und bequem zu verwenden, und bieten ein Interface, das Listen-Operationen, Stack-Operationen, und Array-Operationen unterstützt. Die unterliegende Implementierung verwendet ein *dynamisches Array* von Pointern mit *Overallocation* __[[Quelle]](https://docs.python.org/3/faq/design.html#how-are-lists-implemented-in-cpython)__, d.h. zum einen werden nicht die echten Werte, sondern nur Referenzen auf diese in einer Liste gespeichert, zum zweiten belegen Listen in der Regel mehr Platz, als erwartet, und zum dritten sind Zugriffe über den Index in der Regel immer noch **O**(1), allerdings, wegen der Pointer-Dereferenzierung, langsamer als bei echten Arrays. Auch haben Operationen wie das Ein- oder Ausfügen von Elementen sehr verschiedene  Laufzeiten, je nach dem, wo ein Element im Array steht. Für viele Anwendungen sind Listen in Python eine adäquate Datenstruktur. 

Betrachten Sie folgene Beispiele und probieren Sie sie aus:

In [3]:
lst = ["a", "b", "c", "d", "e"]

i = 0
while i < len(lst):
    print(i, ":", lst[i])
    i = i+1

0 : a
1 : b
2 : c
3 : d
4 : e


Hier wird über die Liste (bzw. das Array) `lst` wie in einer klassischen Programmiersprache  iteriert. Wir haben eine klassische Schleifenvariable, die initialisiert und erhöht wird, und wir indizieren das Array direkt. 

In Python ist man allerdings selten gezwungen, Zählvariablen manuell zu manipulieren. Wesentlich *pythonischer* ist folgende Lösung:

In [4]:
for i in range(len(lst)):
        print(i, ":", lst[i])

0 : a
1 : b
2 : c
3 : d
4 : e


Die Funktion `range()` liefert ein Objekt, dass sich im wesentlichen wie ein Liste von Indices verhält. Mit einem Argument ist das Ergebnis äquivalent zur Liste `[0,...,arg-1]`, liefert also genau die Indices eines Arrays mit `arg` Elementen.

Auch bei dieser Lösung wird noch mit einer expliziten Indexvariable gearbeitet - in vielen klassischen Programmiersprachen ist das nötig. Moderne Sprachen bieten hier oft komfortablere Lösungen - so auch Python, wenn man die Laufvariable nicht braucht:

In [5]:
for l in lst:
    print("Kein expliziter Index!", l)

Kein expliziter Index! a
Kein expliziter Index! b
Kein expliziter Index! c
Kein expliziter Index! d
Kein expliziter Index! e


### Tupel und Dictionaries

*Tupel* werden in Python mit Kommata getrennt und üblicherweise in runde Klammern eingeschlossen: `("a", "b", "c", "d", "e")`. Sie unterscheiden sich funktional von Listen dadurch, dass sie *immutable*, also unveränderlich sind. Weder können Sie Elemente hinzufügen oder löschen, noch können Sie Elemente verändern. Probieren Sie es aus:

In [6]:
lst = ["a", "b", "c", "d", "e"]
lst[1] = "Hallo"
print(lst)

['a', 'Hallo', 'c', 'd', 'e']


In [7]:
tup = ("a", "b", "c", "d", "e")
tup[1] = "Selber hallo!"
print(tup)

TypeError: 'tuple' object does not support item assignment

*Dictionaries* oder *assoziative Arrays* haben in Python dynamische Größe und erlauben die Indizierung über beliebige *immutable* Objekte - Zahlen, Strings, Tupel, aber z.B. keine Listen. Bei Dictionaries müssen die Schlüssel immer angegeben werden. Technisch steht hinter Dictionaries eine *Hash-Table*. Die Geschwindigkeit skaliert ähnlich wie bei echten Arrays, ist aber grundsätzlich ca. 1-2 Größenordnungen langsamer.

Im folgenden Beispiel wird ein Dictionary von Studierenden mit als Index die Matrikelnummer verwendet. Führen Sie die Code-Fragmente der Reihenfolge nach aus.

Auf Elemente (engl. *values*) kann mittels des Indexwert (engl. *key*) zugegriffen werden.

In [8]:
students_by_matr = {"7439876":"Max Muster", "0677231":"Stefan Student", "9238573":"Karla Kommilitona", "6109300":"Alime Absolventin"}
print("Die Matrikelnummer 7439876 hat", students_by_matr["7439876"])

Die Matrikelnummer 7439876 hat Max Muster


Umgekehrt ist die allerdings nicht möglich.

In [9]:
print("Stefan Student hat die Matrikelnummer ", students_by_matr["Stefan Student"])

KeyError: 'Stefan Student'

Neue Einträge können nach der Definition hinzugefügt werden, da das Dictionary nicht *immutable* ist. Dies ist i.d.Regel effizienter als in echten Arrays.

In [10]:
print("Gibt es einen Studi mit Matrikelnummer 4328787?", "4328787" in students_by_matr)
students_by_matr["4328787"] = "Eberhard Erstie"
print("Die Matrikelnummer 4328787 hat", students_by_matr["4328787"])

Gibt es einen Studi mit Matrikelnummer 4328787? False
Die Matrikelnummer 4328787 hat Eberhard Erstie


Ebenso können Einträge entfernt werden.

In [11]:
print("Die Matrikelnummer 6109300 hat", students_by_matr["6109300"])
del students_by_matr["6109300"]
print("Gibt es einen Studi mit Matrikelnummer 6109300?", "6109300" in students_by_matr)

Die Matrikelnummer 6109300 hat Alime Absolventin
Gibt es einen Studi mit Matrikelnummer 6109300? False


Mittels einer for-Schleife können Sie über das Dictionary iterieren.

In [12]:
dct = {1:"a", 2:"b", 3:"c", "vier":"d", (2,3):"e"}
print(dct)
for k in dct:
    print(k)

{1: 'a', 2: 'b', 3: 'c', 'vier': 'd', (2, 3): 'e'}
1
2
3
vier
(2, 3)


Beachten Sie, dass die `for`-Schleife beim Dictionary über die Indexwerte (oder *Schlüssel*) iteriert. Um an die Werte zu kommen, muss man explizit dereferenzieren:

In [13]:
for k in dct:
    print(k, ":", dct[k])

1 : a
2 : b
3 : c
vier : d
(2, 3) : e


### Python Arrays (als allgemeine Arrays nicht empfohlen!)

Python implementiert "echte" Arrays im Sinn von C im Modul `array`. Allerdings ist dieser 
Typ vor allem auf den effizienten Datenaustausch mit Betriebssystemfunktionen und externen
Datenquellen- und Senken ausgelegt. Insbesondere sind diese Arrays als C-Arrays und mit den
Beschränkungen von C-Datentypen angelegt. Entsprechend werden nur wenige Python-Datentypen
unterstützt, und diese nur beschränkt. Näheres finden Sie natürlich 
__[im Manual](https://docs.python.org/3/library/array.html)__.


### NumPy Arrays

Python as *die* Sprache im Bereich Datascience muss natürlich den effizienten Umgang mit 
Arrays unterstützen. Es tut dies mit der Erweiterung NumPy, die heute als Modul `numpy` in
praktisch jeder Python-Installation enthalten ist und ein sehr mächtiges Array-Konzept 
unterstützt. NumPy ist speziell dafür entwickelt, numerische Programmierung in Python
einfacher und effizienter zu machen. NumPy und SciPy stecken in den meisten komplexen
wissenschaftlichen Programmen, die in Python entwickelt werden. Die 
__[volle Dokumentation](https://numpy.org/doc/stable/index.html)__ zu NumPy finden Sie 
online, eine __[Einführung in NumPy Arrays](https://numpy.org/doc/stable/user/absolute_beginners.html)__ 
ebenfalls.

NumPy Arrays haben im Vergleich zu Python-Listen einige Einschränkungen,
sind im Gegenzug aber für viele Operationen effizienter und unterstützen auch einen reicheren
Vorrat an Methode. Insbesondere:
- In NumPy-Arrays sind alle Elemente vom gleichen Typ. Diesen ermittelt NumPy in der Regel automatisch
  wenn das Array angelegt wird (er kann aber auch vom Programmierer vorgegeben werden), und er
  kann danach nicht verändert werden.
- NumPy-Arrays sind n-dimensional (die häufigsten Spezialfälle sind 1-dimensional - also
  von der Struktur wie Vektoren, und 2-dimensional - analog zu Matrizen). Deswegen heißt die
  Klasse, die sie implementiert, auch `ndarray`.
- NumPy-Arrays sind in allen Dimensionen konstant groß. Ein 2-D-Array hat also die Form einer
  rechteckigen Tabelle, ein 3-D-Array die Form eine Quaders von aufeinander gestapelten Tabellen,
  und so weiter.
- Die Gesamtgröße eines NumPy-Arrays kann nach Erschaffung nicht mehr verändert werden. Insbesondere
  funktionieren Methoden wie `pop()` oder `append()` nicht. Man kann die (gedachte) Anordnung
  der Elemente des Arrays verändern, solange sich die Gesamtgröße nicht ändert. Ein Array der
  Größe 4x6 kann also auch als 2x12, 6x4, 2x3x4 oder sogar als 4-d-Array 2x3x2x2 gedacht werden.
  Dabei verändert sich die Lage der Daten im Speicher nicht, nur die *Form* (*Shape*) des
  Arrays.
- Slicing (das definieren von Sub-Arrays) funktioniert anders als mit klassischen Python-Listen.
  Während bei Listen ein Befehl wie `a[2:4]` eine Kopie der Teilliste zurückgibt, wird für ein
  `ndarray` ein *View* auf einen Ausschnitt des Originalarrays erzeugt. Der Unterschied wird
  offensichtlich, wenn man eines von beiden verändert - dabei ändert sich immer das Gegenstück
  mit.

Um NumPy nutzen zu können, müssen Sie zunächst das Modul importieren. Traditionell wird
NumPy unter dem Namen `np` importiert, um bequemen Zugang zu allen Klassen und 
Funktionen zu haben, ohne alle Namen direkt in den globalen Namespace zu importieren:
  
  

In [1]:
import numpy as np

a = np.array([[1, 2, 3],[4, 5, 6]])

print(a)


[[1 2 3]
 [4 5 6]]


Damit ist ein NumPy-Array mit zwei Zeilen und drei Spalten angelegt. 
<div class="aufgabe">
    <h3>Erste Experimente mit ndarray</h3>
Probieren Sie die folgenden
Befehle aus. Können Sie die Ergebnisse erklären?
</div>

In [ ]:
print(a[1][2])
print(a[1, 2])
print(a[1, 4])

**Erklärung**

---
<span style="color:red">Student Answer</span> <br>
[1][2] -> 6: Element 1 im großen Array, Element 2 im kleineren Array <br>
[1, 2] -> Alternative Schreibweise <br>
[1, 4] -> Element 1 hat keine 4 Elemente

---



In [15]:
a[1,1]=18
print(a)

[[ 1  2  3]
 [ 4 18  6]]


**Erklärung**

---
## <span style="color:red">Student Answer</span>

*Double-click and add your answer between the lines*

---

In [16]:
a[1,1]=3.1415
print(a)
a[1,1]="Hallo"
print(a)

[[1 2 3]
 [4 3 6]]


ValueError: invalid literal for int() with base 10: 'Hallo'

**Erklärung**

---
## <span style="color:red">Student Answer</span>

*Double-click and add your answer between the lines*

---

In [2]:
a[1,1] = len(a)
print(a)

[[1 2 3]
 [4 2 6]]


In [3]:
b = [len]
print(b)

[<built-in function len>]


**Erklärung**

---
## <span style="color:red">Student Answer</span>

*Double-click and add your answer between the lines*

---

In [4]:
l = [[1, 2, 3],[4, 5, 6]]
a = np.array([[1, 2, 3],[4, 5, 6]])
print(l)
print(a)
print()

l1 = l[1][0:2] # 0 bis ausschließlich 2
a1 = a[1,0:2]
print(l1)
print(a1)
print()

l1[1] = 17
a1[1] = 17 # Keine Kopie, sondern ein "View" von a -> Veränderung verändert auch a!
print(l)
print(a)
print()

a[1,0]=32
print(a1)
print()

[[1, 2, 3], [4, 5, 6]]
[[1 2 3]
 [4 5 6]]

[4, 5]
[4 5]

[[1, 2, 3], [4, 5, 6]]
[[ 1  2  3]
 [ 4 17  6]]

[32 17]



**Erklärung**

---
## <span style="color:red">Student Answer</span>

*Double-click and add your answer between the lines*

---

#### Dimensionen und Größen

Wie die meisten Objekte in Python sind auch Objekte vom Typ `ndarray` zu einem guten Teil selbstbeschreibend,
d.h. man kann viele ihrer Eigenschaften über Member-Variablen oder Methoden erfragen. Im Fall eines `ndarray`
mit Namen `a` sind das insbesondere:

- `a.ndim`: Anzahl der Dimensionen des Arrays. Das Ergebnis ist 1 für einfache 1-dimensionale Arrays,
  zwei für tabellenartige 2d-Arrays, und so weiter
- `a.shape`: Größe des Arrays in der jeweiligen Dimension. Das Ergebnis ist ein Tupel von `a.ndim` vielen Integer-Werten.
- `a.size`: Gesamtanzahl der Elemente des Arrays. Der Wert ist das Produkt der einzelnen Dimensionsgrößen.
- `a.dtype`: Typ der Arrayelemente - das ist für die meisten unserer Anwendungen weniger wichtig.

Probieren Sie es einmal aus:

In [20]:
a = np.array([[1, 2, 3],[4, 5, 6]])
b = np.array([1, 2, 3, 4, 5, 6 ])
c = np.array([[1, 2, 3, 4, 5, 6 ]])

print("a.ndim:", a.ndim)
print(a)
print(a.shape)
print(a.size)
print()

print("b.ndim:", b.ndim)
print(b)
print(b.shape)
print(b.size)
print()

print("c.ndim:", c.ndim)
print(c)
print(c.shape)
print(c.size)
print()


a.ndim: 2
[[1 2 3]
 [4 5 6]]
(2, 3)
6

b.ndim: 1
[1 2 3 4 5 6]
(6,)
6

c.ndim: 2
[[1 2 3 4 5 6]]
(1, 6)
6



<div class="aufgabe">
    <h3>Ergebnisse interpretieren</h3>
Können Sie die Ergebnisse erklären?
</div>
    
**Erklärung**

---
## <span style="color:red">Student Answer</span>

*Double-click and add your answer between the lines*

---
  
#### Indexing für Fortgeschrittene

Eine Besonderheit von `ndarray` ist, dass nicht nur einzelne Elemente über die Koordinaten angesprochen 
werden können, sondern auch ganze Gruppen von Elementen, und zwar sowohl nach Lage im Array, als auch
semantisch über den Inhalt der Felder. Ersteres haben wir ja bereits in obigem Beispiel gesehen: Man kann statt
eines festen Index-Wertes für jede Dimension immer auch einen Bereich oder *Slice* der Form `X:Y` angeben
und damit die Werte mit den Indices `X` bis `Y-1` selektieren (beide können auch weggelassen werden, dann
geht es ab 0 bzw. bis zum Ende des Wertebereichs). Wichtig: Dabei werden in der Regel *Views* zurückgegeben,
also Ansichten auf das existierende Array. Wenn sie genau wissen wollen, wann *Views* und wann Kopien erzeugt
werden, dann __[hilft ihnen das Manual](https://numpy.org/doc/stable/user/basics.copies.html)__.



In [5]:
a = np.array([[1, 2, 3],[4, 5, 6],[7, 8, 9]])
print(a[0:2,1:3])


[[2 3]
 [5 6]]


Tatsächlich ist die Slice-Notation bei NumPy-Arrays noch vielseitiger: Wir können auch noch eine Schrittweite
angeben, also einen Slice der Form `X:Y:Z` verwenden. Zu lesen ist das als "Nimm ab Index `X` bis (ausschließlich) 
Index `Y` jeden `Z`-ten Wert:

In [25]:
a = np.array(range(30))
print(a)
print(a[10:25:3])

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]
[10 13 16 19 22]


Das funktioniert auch im mehrdimensionalen Fall. Was ergibt folgender Code?

In [26]:
b = np.array([range(10), range(0,20,2), range(10,20)])
print(b)
print(b[:,5:15:2]) # Zu viele Elemente werden akzeptiert, aber irreführend (lieber 10 als 15)


[[ 0  1  2  3  4  5  6  7  8  9]
 [ 0  2  4  6  8 10 12 14 16 18]
 [10 11 12 13 14 15 16 17 18 19]]
[[ 5  7  9]
 [10 14 18]
 [15 17 19]]


Tatsächlich kann man solchen Auszügen aus einem Array auch Werte zuweisen, wenn man einen
Wert kompatibler Struktur wählt:

In [29]:
a = np.array(range(30))
print(a)
a[10:25:3] = [666,666,666,666,666]
print(a)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]
[  0   1   2   3   4   5   6   7   8   9 666  11  12 666  14  15 666  17
  18 666  20  21 666  23  24  25  26  27  28  29]


...und kompatibel sind recht viele Werte, wie im nächsten Abschnitt angerissen wird.
Doch zuvor wollen wir uns noch kurz mit der semantischen Addressierung beschäftigen. So
können wir aus einem Array z.B. alle Werte extrahieren, die bestimmte Eigenschaften
haben. Dazu wird die Bedingung in die eckigen Klammern gesetzt, z.B.: `a[a%2==0]`.
Probieren Sie es aus!

In [30]:
print("a:", a)
print(a[a%2==0])
print()
print("b:", b)
print(b[b>10])

a: [  0   1   2   3   4   5   6   7   8   9 666  11  12 666  14  15 666  17
  18 666  20  21 666  23  24  25  26  27  28  29]
[  0   2   4   6   8 666  12 666  14 666  18 666  20 666  24  26  28]

b: [[ 0  1  2  3  4  5  6  7  8  9]
 [ 0  2  4  6  8 10 12 14 16 18]
 [10 11 12 13 14 15 16 17 18 19]]
[12 14 16 18 11 12 13 14 15 16 17 18 19]


Beachten Sie: In dem Fall ist auch der Auszug aus dem zweidimensionalen Array `b` ein 1d-Array.
Versuchen Sie einmal, diesem Auszug etwas zuzuweisen!

In [31]:
b[b>10] = 1
print(b)

[[ 0  1  2  3  4  5  6  7  8  9]
 [ 0  2  4  6  8 10  1  1  1  1]
 [10  1  1  1  1  1  1  1  1  1]]


Python und NumPyu überraschen hier klassische Informatiker mit einer
konsequent umgesetzten Technik, die man so nicht erwarten würde. Aber warum
konnten wie dem relativ langen Auszug aus dem Array den einfachen Wert
1 zuweisen, der dann für alle Werte eingesetzt wurde?



#### Broadcasting

*Broadcasting* ist eine Technik, mit der NumPy Operationen zwischen Arrays kompatibler
Größe oder zwischen Arrays und Einzelwerten (Skalaren) sehr effizient ermöglicht. 
Grundsätzlich erfolgen Operationen zwischen Arrays elementweise, wobei die jeweils
korrespondierenden Werte interagieren:

In [38]:
a = np.array([1, 2, 3, 4])
b = np.array([8, 7, 6, 5])
print(a+b)
print(a*b)
print()

c = np.array(range(20))
c[8:12] = (a+b)
print(c)

[9 9 9 9]
[ 8 14 18 20]

[ 0  1  2  3  4  5  6  7  9  9  9  9 12 13 14 15 16 17 18 19]


__[Broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html)__ ermöglicht es, einen 
passenden "kleineren" Datentyp über den "größeren" Partner zu verteilen. Den einfachsten Anwendungsfall
haben Sie bereits gesehen: Wir können Operationen zwischen einem Array und einem Einzelwert 
(oder *Skalar*) durchführen. Dabei wird der Skalar so behandelt, als wäre er ein Array passender
Größe, bei dem alle Felder mit dem Wert des Skalars besetzt sind.

Das funktioniert wunderbar auch in komplexeren Fällen, z.B. wenn man ein Array mit den Dimensionen
3x4 mit einem Array mit den Dimensionen 4x1 verknüpft. Dabei wird das zweite Array virtuell 3 mal 
dupliziert, so dass die Größen passen:

In [39]:
a = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 9, 9, 9]])
b = np.array([11, 12, 13,14])
print(a.shape, b.shape)
print(a+b)

(3, 4) (4,)
[[12 14 16 18]
 [16 18 20 22]
 [20 21 22 23]]


<div class="aufgabe">
    <img src="images/Eratosthenes_profile.png" width=120 align=right alt="Erathosthenes betrachtet die Aufgabenstellung ;-)"/>
    <h3>Erathostenes mal anders</h3>    
    Erinnern Sie sich an das <em>Sieb des Erathostenes</em> aus der ersten Veranstaltung? Schreiben Sie nun eine 
    Funktion <tt>erathosethenes_sieve(max_val)</tt>, die das Sieb des Erathosthenes noch einmal implementiert
    und die Primzahlen bis (einschließlich) <tt>max_val</tt> als Liste zurückgibt.
    Verwenden Sie für das eigentliche Sieb ein NumPy-Array. Sie sollten mit einer einzigen expliziten
    Schleife auskommen können!
</div>
<details class="hint">
<summary></summary>
   Nehmen Sie für das eigentliche Sieb ein Array von natürlichen Zahlen, 
   nicht von Boolschen Werten, und markieren Sie "durchgestrichene" Werte, 
   indem Sie sie auf 0 setzen.
    <details class="nexthint">
    <summary></summary>
       Sie können alle Vielfachen einer Zahl durch einen Broadcast über einen geeigneten Slice erreichen.
    </details>
</details>


In [87]:
import numpy as np

# YOUR CODE HERE
def erathosethenes_sieve(max_val):
    primes = []
    a = np.array(range(max_val+1))
    # a = a[1:max_val]

    a[1] = 0

    for p in a:
        if p != 0:
            primes.append(p)
            a[2*p::p] = 0
        # min_val = a[a != 0].min()
        # primes.append(min_val)
        # a[(a % min_val == 0)] = 0

    return primes


print(erathosethenes_sieve(1000))
# Erwartetes Ergebnis: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53]


[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]


## Vektoren

In der Physik sind Vektoren Objekte, die durch eine Länge und eine
Richtung beschrieben werden. In der Mathematik sind Vektoren Elemente
eines Vektorraums. Ein Vektorraum ist eine mathematische Struktur,
konkret eine abelsche Gruppe (von "Vektoren") über einem Körper K (von
"Skalaren"), bei der eine zusätzliche Verknüpfung (die
Skalarmultiplikation) das Verknüpfen von Vektoren und Skalaren
ermöglicht. Für diese Skalarmultiplikation werden zusätzlich
Distributivgesetze und Assoziativgesetz gefordert.

In dieser Vorlesung bewegen wir uns zwischen diesen beiden
Definitionen. Für uns sind Vektoren Elemente von $\mathbb{R}^n$, also $n$-Tupel
von reelen Zahlen. Dabei ist $n$ fest - d.h. jeder gegebene Vektorraum
enthält nur Vektoren mit $n$ Elementen.  Die reelen Zahlen $\mathbb{R}$ selbst 
bilden den *Skalarkörper*, und sind die zugehörigen *Skalare*. Wenn Ihnen
das Wort komisch vorkommt: Ein Skalar *skaliert* einen Vektor - wenn 
sie einen Vektor mit dem Skalar 2 multiplizieren, so behält er seine 
Richtung, wird aber doppelt so lang.

### Grundlagen 

<div class="definition">
   <h3>$\mathbb{R}^n$ als Vektorraum</h3>
    <p>
    Wir setzen die reelen Zahlen $\mathbb{R}$ voraus, die mit den üblichen Operationen 
    $+:\mathbb{R}\times\mathbb{R}\to\mathbb{R}$ und
    $\cdot:\mathbb{R}\times\mathbb{R}\to\mathbb{R}$ einen <em>algebraischen Körper</em>
    bilden. Für $x\in\mathbb{R}$ bezeichnet $-x$ das <em>inverse Element</em> zu $x$ (also 
    die Zahl, die sich mit $x$ zu 0 addiert), und $x-y$ steht kurz für $x+(-y)$.
    </p>
    <p>
    Sei $n\in \mathbb{N}, n>0$. Die Menge $\mathbb{R}^n$ ist die Menge aller <em>n-Tupel über 
    $\mathbb{R}$</em></em>, also von allen Tupeln der Form $\langle x_1, \ldots, x_n \rangle$, 
    wobei $x_i \in \mathbb{R}$ für alle $i$ von $1$ bis $n$ gilt. 
    Wir definieren folgende Operationen:
    <ul>
    <li>$\oplus:\mathbb{R}^n \times \mathbb{R}^n \to \mathbb{R}^n, 
        \langle{} x_1, \ldots, x_n \rangle{} \oplus \langle y_1, \ldots, y_n \rangle \mapsto \langle x_1+y_1, \ldots, x_n+y_n \rangle$
        (die <em>Vektoraddition</em>, also die Addition zweier Vektoren zu einem dritten)
    </li>
    <li>$\otimes:\mathbb{R} \times \mathbb{R}^n \to \mathbb{R}^n, 
        \lambda \otimes \langle y_1, \ldots, y_n \rangle \mapsto \langle \lambda \cdot y_1, \ldots, \lambda \cdot y_n \rangle$ 
        (die <em>Skalarmultiplikation</em>, d.h. die Multiplikation einer reelen Zahl mit einem Vektor zu einem neuen Vektor)
    </li>
    </ul>
    </p>
    <p>
    Damit ist $(\mathbb{R}^n, \oplus, \otimes)$ ein <em>Vektorraum</em>. Die 
    Elemente von $\mathbb{R}^n$ heißen <em>Vektoren</em>.</p>
</div>


Vektoren im $\mathbb{R}^n$ sind also (mathematische) Tupel. Davon zu unterscheiden
ist die Python-Datenstruktur der Tupel. Wir können Python-Tupel verwenden, um
Vektoren zu repräsentieren, aber das ist nur die halbe Miete - die Operationen
fehlen ja. Später werden wir eine Klasse `Vector` definieren, die Vektoren als
`ndarray`s darstellt, und die Vektor-Operationen als Klassenmethoden definiert.

Einfache Spezialfälle sind $\mathbb{R}^2$ - dann reden wir über Vektoren in der
Ebene, und $\mathbb{R}^3$ - Vektoren im 3-dimensionalen Raum. Diese beiden Fälle
haben eine Vielzahl von Anwendungen in Technik und Physik.

Zur Verwendung von Vektoren in mathematischen Formeln gibt es verschiedene 
Konventionen. Entweder werden Variablen, die Vektoren repräsentieren, 
**fett** gedruckt. Alternativ werden die Variablennamen auch mit einem Pfeil
über dem Namen dargestellt: $\vec{a}$. Wir verwenden hier die erste 
Schreibweise. Zur einfacheren Schreibweise wird auch oft auf die Unterscheidung
zwischen der Skalaraddition $+$ und der Vektoradition $\oplus$ verzichtet, wenn
die Typen klar sind. Gleiches gilt für $\otimes$ und $\cdot$, wobei beide
Multiplikationszeichen auch oft weggelassen werden. Also: 
$\lambda (\mathbf{a}+\mathbf{b})$ ist die vereinfachte Schreibweise 
für $\lambda \otimes (\mathbf{a}\oplus\mathbf{b})$.

<div class="aufgabe">
<h3>Vektorrechnung</h3>
    Seien $\mathbf{a}=\langle 1, 2, 3 \rangle, \mathbf{b}=\langle 0, 2, 4 
    \rangle, \mathbf{n} = \langle 0, 0, 0 \rangle$. Berechnen Sie die folgenden 
    Ausdrücke:
    <ul>
    <li>$\mathbf{a}+\mathbf{b}$</li>
    <li>$\mathbf{a}+3\mathbf{b}$</li>
    <li>$\mathbf{b}+3\mathbf{n}$</li>
    <li>$\mathbf{b}-3\mathbf{a}$</li>
    <li>$2\mathbf{a}$</li>
    <li>$\mathbf{a}+\mathbf{a}$</li>
    </ul>
</div>

Lösung:
---
## <span style="color:red">Student Answer</span>

*Double-click and add your answer between the lines*

---

<div class="remark">
    <img src="images/Sir_Isaac_Newton_1702.jpg" width=120 align=right alt="Image of Isaac Newton" />   
    <h3>Philosophiæ Naturalis Principia Mathematica</h3>
    <p>
    Newtons berühmte Formel $F=m\cdot{}a$ (Kraft = Masse mal Beschleunigung) kann man
    auch als Vektorgleichung schreiben: $\mathbf{F}=m \cdot{} \mathbf{a}$. Kraft und 
    Beschleunigung sind Größen, denen man auch eine Richtung zuordnen kann, d.h. es sind
    <em>vektorielle</em> Größen im $\mathbb{R}^3$. Die Masse $m$ ist dagegen ein reiner Skalar. 
    </p>
    <p>
    Wenn man direkt mit Vektoren rechnet, werden viele Problem der Physik einfacher und
    vor allem eleganter, als wenn man die Größe und die Richtung verschiedener Konzepte
    (Kräfte, Beschleunigungen, Bewegungen, ...) jeweils separat berechnen muss.
    </p>
</div>

Wenn wir ein karthesisches Koordinatensystem annehmen, dann
repräsentiert der Vektor <1, 2, 3> die Verschiebung eines Punktes um
eine Einheit in der X-Koordinate, zwei Einheiten in der Y-Koordinate,
und drei Einheiten in der Z-Koordinate.

Analog können Vektoren aber auch selbst als Punkte in Raum
interpretiert werden. Der Übergang zwischen beiden Sichtweisen ist
einfach - der Vektor entspricht dem Punkt, auf den er den
Koordinatenursprung verschieben würde. 

Im karthesischen Koordinatenraum können wir auch verschiedene Abstandsmaße
definieren. Das wichtigste beruht auf dem *Skalarprodukt*, einer 
Verknüpfung zweier Vektoren zu einem Skalar (in unserem Fall also einer
reelen Zahl). 
<div class="definition">
   <h3>Skalarprodukt im $\mathbb{R}^n$</h3>
    Seien $\mathbf{a}=\langle a_1, \ldots, a_n \rangle, \mathbf{b}=\langle b_1, \ldots, b_n \rangle \in \mathbb{R}^n$. Das
    Skalarprodukt der beiden Vektoren ist $\mathbf{a}\cdot \mathbf{b} = \sum_{i=1}^n a_ib_i$.
</div>

Das Skalarprodukt ist also die Summe der Produkte der korrespondierenden Vektorelemente.
Über das Skalarprodukt kann man einem Vektor auch eine Länge oder *Norm* zuordnen:
<div class="definition">
   <h3>Euklidsche Norm</h3>
    Seien $\mathbf{a} \in \mathbb{R}^n$. Die Norm von $\mathbf{a}$ ist $\|\mathbf{a}\|=\sqrt{\mathbf{a}\cdot \mathbf{a}}$.
</div>

Die so aus dem Skalarprodukt definierte Norm entspricht exakt der geometrische Länge
eines Vektors, die man auch durch iterierte Anwendung des Satzes des Pythagoras
berechenen kann.

<div class="remark">
    <img src="images/clouds.png" width=120 align="center" alt="Clouds with lightning" />   
    <h3>Skalarmultiplikation vs. Skalarprodukt</h3>
        <ul>
        <li>Die <em>Skalarmultiplikation</em> ist eine Operation zwischen einem Skalar und einem Vektor. 
             Sie wird auch "äußeres Produkt" genannt, weil einer der Operatoren kein Element des
             Vektorraums  ist (sondern des "äußeren" Skalarköpers - daher der Name). Das Ergebnis ist wieder 
            ein Vektor. Jeder Vektorraum muss eine Skalarmultiplikation haben - sie ist eine der definierenden 
             Eigenschaften.</li>
        <li>Viele, aber nicht alle Vektorräume definieren zusätzlich ein sogenanntes <em>inneres Produkt</em>,
            das <em>Skalarprodukt</em>. Dieses verknüpft zwei Vektoren, das Ergebnis ist ein Skalar (daher hier der Name).
            Vektorräume mit Skalarprodukt werden auch <em>Prähilberträume</em> oder <em>euklidsche Räume</em>
            genannt - leider ist die Nomenklatur aber nicht einheitlich. Unabhängig vom Wert von $n$ ist 
            $\mathbb{R}^n$ ein euklidscher (Vektor-)Raum, und hat ein Skalarprodukt.
        </li>
    </ul>
    Diese beiden Begriffe sollte man nicht verwechseln
</div>

<div class="aufgabe">
<h3>Skalarprodukt und Norm</h3>
    Seien $\mathbf{a}=\langle 1, 2, 3 \rangle, \mathbf{b}=\langle 0, 2, -4 \rangle, 
    \mathbf{c}=\langle 0, 0, 1 \rangle,\mathbf{d} = \langle 0, 1, 0 \rangle$ und 
    $\mathbf{o} = \langle 0, 0, 0 \rangle$. Berechnen Sie die folgenden 
    Ausdrücke:
    <ul>
    <li>$\mathbf{a}\cdot\mathbf{b}$</li>
    <li>$\mathbf{a}\cdot\mathbf{c}$</li>
    <li>$\mathbf{a}\cdot\mathbf{d}$</li>
    <li>$\mathbf{c}\cdot\mathbf{d}$</li>
    <li>$\mathbf{a}\cdot\mathbf{o}$</li>
    <li>$\|\mathbf{a}\|$</li>
    <li>$\|\mathbf{b}\|$</li>
    <li>$\|\mathbf{c}\|$</li>
    <li>$\|\mathbf{d}\|$</li>
    <li>$\|\mathbf{o}\|$</li>
    <li>$\|\mathbf{a}-\mathbf{b}\|$</li>
    </ul>
</div>

Lösung:

---
## <span style="color:red">Student Answer</span>

*Double-click and add your answer between the lines*

---

### Einschub: Vektoren in der KI

In der künstlichen Intelligenz treffen wir Vektoren vor allem im Gebiet des maschinellen
Lernen an. Objekte, die wir z.B. klassifizieren wollen, können als *Feature-Vektoren*
kodiert werden. Im Prinzip können einzelne Features aus verschiedenen Domänen kommen,
in der Praxis gehen viele Lernverfahren (z.B. klassische neuronale Netze) aber besser mit 
rein numerischen Vektoren um - damit liegt die Repräsentation eines Objektes wieder im
$\mathbb{R}^n$ und wir sind im Bereich der klassischen linearen Algebra.

Schauen wir uns als Beispiel eine Reihe von Sportlern an:

|Name             |Größe (cm)|Gewicht (kg)|Sportart|
|---              | --- | ---    |---     |
|Usain Bolt       | 195 | 94     | Sprint  |
|Eliud Kipchoge   | 167 | 52     | Marathon|
|Stephen Kiprotich| 172 | 56     | Marathon|
|Bam Bam Bigelow  | 193 | 177    | Pro-Wrestling|
|Hulk Hogan       | 201 | 137    | Pro-Wrestling|
|Magic Johnson    | 206 | 100    | Basketball|
|Dirk Nowitzki    | 213 | 111    | Basketball|
|LeBron James     | 206 | 113    | Basketball|
|Shaquille O'Neal | 216 | 147    | Basketball|
|Michael Jordan   | 198 |  98    | Basketball|
|Big Van Vader    | 196 | 204    | Pro-Wrestling|
|Sting            | 188 | 113    | Pro-Wrestling|
|Jesse Owens      | 180 |  75    | Sprint |
|Carl Lewis       | 188 |  80    | Sprint |
|Ben Johnson      | 177 |  75    | Sprint |
|Larry Bird       | 206 | 100    | Basketball|
|Kelvin Kiptum    | 189 | 65     | Marathon|

<img src="images/sportler_knn.png" width=600 align=right alt="Diagramm der Sportler und ihrer Eigenschaften"/>

Die Klassifikationsaufgabe besteht darin, aus Größe und Gewicht die Sportart
zu prognostizieren. Am einfachsten geht das vielleicht mit dem 
*Nearest Neighbour*-Verfahren. Dabei sucht man den ähnlichsten bekannten
Sportler, und prognostiziert dessen Sportart. 

Eine Variante ist das $k$NN-Verfahren,
wo man nicht einen, sondern die $k$ nächsten Nachbarn bestimmt, und die häufigste
Klasse unter diesen bestimmt. So oder so braucht man dafür aber *Abstandsmaße* oder
Ähnlichkeitsmaße.

Eine andere interessante Aufgabe wäre es hier, die unterschiedlichen 
Klassen ohne Vorkenntnisse zu erkennen - sogenanntes "unsupervised
learning". Dabei würde z.B. der __[*K-Means*-Algorithmus](https://en.wikipedia.org/wiki/K-means_clustering)__ 
verwendet, um Cluster von ähnlichen Individuen zu erkennen. Auch dabei
kommen Abstandsmaße zum tragen.

Übrigens: Auch aktuelle KI-Systeme wie z.B. *Large Language Models* basieren weitgehend
auf numerischen Vektoren und einer großen Menge linerarer Algebra, gewürzt mit kleinen
aber kritischen Anteilen von Analysis.


### Abstandsmaße und Ähnlichkeiten

Wenn wir Vektoren als Punkte im karthesischen Raum interpretieren, dann ist der natürlichste "Abstand" 
der beiden Vektoren die Länge der Linie, die die beiden Punkte verbindet. Dieses Abstandsmaß nennen
wir auch den *Euklidschen Abstand*, und wir können ihn elegant über bereits bekannte Konzepte 
definieren - er ergibt sich nämlich gerade als die Länge (oder "euklidsche Norm") des Differenzvektors.

<div class="definition">
   <h3>Euklidscher Abstand</h3>
    <p>
    Seien $\mathbf{a}=\langle a_1, \ldots, a_n \rangle, \mathbf{b}=\langle b_1, \ldots, b_n \rangle \in \mathbb{R}^n$.
    Der euklidsche Abstand $d_{euklid}:\mathbb{R}^n\times \mathbb{R}^n \to \mathbb{R}$ ist definiert als 
    $d_{euklid}(\mathbf{a},\mathbf{b}) = \|\mathbf{a}-\mathbf{b}\|$.    
    </p>
    <p>Direkt berechnen kann man ihn auch als $d_{euklid}(\mathbf{a},\mathbf{b}) = \sqrt{\sum_{i=1}^n (a_i-b_i)^2}$.</p>
</div>

Ein alternatives Abstandsmaß ist der sogenannte "Manhattan-Abstand". Dieser ist inspiriert von dem
rechtwinkligen Straßennetz in Manhattan, bei der der kürzeste praktisch realisierbare Weg immer
den Straßen folgen muss, und die Abkürzung über die Luftlinie (die dem euklidschen Abstand entspricht)
nicht möglich ist. Bei der sogenannten Manhattan-Metrik muss der Weg also immer parallel zu den
Koordinaten-Achsen gemessen werden - also für jede Dimension einzeln:

<div class="definition">
   <h3>Manhattan-Abstand</h3>
    <p>
    Seien $\mathbf{a}=\langle a_1, \ldots, a_n \rangle, \mathbf{b}=\langle b_1, \ldots, b_n \rangle \in \mathbb{R}^n$.
    Der Manhattan-Abstand $d_{manhattan}:\mathbb{R}^n\times \mathbb{R}^n \to \mathbb{R}$ ist definiert als $d_{manhattan}(\mathbf{a},\mathbf{b}) = \sum_{i=1}^n (|a_i-b_i|)$.
    </p>    
</div>

Sowohl der euklidsche Abstand als auch der Manhattan-Abstand sind echte *Metriken*, d.h. die Abstände verschiedener Punkte
sind immer positiv, und es gilt die *Dreiecksungleichung*: Die Summe der Abstände zwischen **a**/**b** und **b**/**c** 
ist immer mindestens so groß wie der direkte Abstand **a**/**c**. Wir betrachten zwei Punkte als *ähnlich*, wenn sie im
Raum nahe zusammenliegen - je näher, je ähnlicher. Auf dieser Annahme beruht z.B. das *Nearest-Neighbor*-Verfahren
zur Klassifikation von Objekten. 

Manchmal sind aber auch andere Ähnlichkeitsmaße nützlich. Ein Beispiel dafür ist die 
Kosinus-Ähnlichkeit. Sie betrachtet nicht die Lage der durch die Vektor-Koordinaten beschriebenen 
Punkte im Raum, sondern die *Richtung* der Vektoren. Konkret wird der Kosinus des Winkels zwischen
zwei Vektoren berechnet. Zwei Vektoren, die in exakt die gleich Richtung zeigen, also parallel 
zueinander sind, umschließen einen Winkel von 0°, und haben somit eine Ähnlichkeit von 1. Vektoren,
die im rechten Winkel zueinander stehen, haben eine Ähnlichkeit von 0, und Vektoren, die *antiparallel*
sind, haben eine Ähnlichkeit von -1. Da ein Vektor der Länge 0 keine Richtung hat, kann man zu ihm
natürlich auch keine Kosinus-Ähnlichkeit bestimmen.

Formal definiert ist die Kosinus-Ähnlichkeit wie folgt:

<div class="definition">
   <h3>Kosinus-Ähnlichkeit</h3>
    <p>
    Seien $\mathbf{a}=\langle a_1, \ldots, a_n \rangle, \mathbf{b}=\langle b_1, \ldots, b_n \rangle \in \mathbb{R}^n$ zwei
    Vektoren mit Länge größer 0.
    Die Kosinus-Ähnlichkeit $s_{cosine}:\mathbb{R}^n\times \mathbb{R}^n \to [-1,1]$ ist definiert als $s_{cosine}(\mathbf{a},\mathbf{b}) = \frac{\mathbf{a}\cdot\mathbf{b}}{\|\mathbf{a}\|\|\mathbf{b}\|}$.
    </p>    
    <p>
    Sie berechnet sich also als das Skalarprodukt der beiden Vektoren, geteilt durch das Produkt der Längen der Vektoren.
    </p>
</div>

Eine Anwendung der Kosinus-Ähnlichkeit ist z.B. beim Dokumenten-Retrieval oder die Dokumenten-Zuordnung. Dabei wird der *Häufigkeitsvektor* für die Wortverteilung in einem Dokument bestimmt. Dokumente mit ähnlichen Inhalten haben typischerweise
ähnliche Wortverteilungen, und verschiedene Autoren haben verschiedene Vorlieben beim Vokabular. Da nur die Richtung der Vektoren berücksichtigt wird, spielt die Länge der Dokumente (und damit der Vektoren) bei der Kosinus-Ähnlichkeit keine Rolle.

<div class="remark">
    <h3>Abstand und Ähnlichkeit illustriert</h3>
    <img src="images/vector_metrics_cropped.png" width=600 align=right alt="Diagram zur Illustration von Euklidscher Metrik, Manhattan-Metrik und Kosinus-Ähnlichkeit"/>
    Das nebenstehende Diagramm zeigt die Vektoren <3,9> und <9,4> in der 
    euklidschen Ebene und die verschiedenen Abstands- und Ähnlichkeitsmaße.
    <ul>
        <li>Der kürzeste Abstand (die euklidsche Distanz) ergibt sich als die Hypothenuse
            eines rechtwinkligen Dreiecks mit Katheten der Läge 6 und 5, also die Wurzel
            aus 25+36=61. Im Diagramm ist der Abstand als rote Linie markiert.</li>
        <li>Die Manhattan-Distanz kann man einfach entlang einer beliebigen Zick-Zack-Linie
            vom Endpunkt von $v_1$ zum Endpunkt von $v_2$, die sich nie von letzterem entfernt,
            auszählen, z.B. entlang der grünen Linie. Das Ergebnis ist 6+5=11. </li>
        <li>Für die Kosinus-Ähnlichkeit berechnen wir erst das Skalarprodukt
            der beiden Vektoren (63) und teilen das Ergebnis durch die beiden Längen
            der Vektoren.</li>
    </ul>
</div>





<div class="aufgabe">
<h3>Abstand und Ähnlichkeit</h3>
    Seien $\mathbf{a}=\langle 1, 2, 3 \rangle, \mathbf{b}=\langle 0, 2, -4 \rangle, 
    \mathbf{c}=\langle 0, 0, 1 \rangle,\mathbf{d} = \langle 0, 1, 0 \rangle$ und 
    $\mathbf{o} = \langle 0, 0, 0 \rangle$. Berechnen Sie die folgenden 
    Ausdrücke:
    <ul>
    <li>$d_{euklid}(\mathbf{a}, \mathbf{b})$</li>
    <li>$d_{manhattan}(\mathbf{a}, \mathbf{b})$</li>
    <li>$s_{cosine}(\mathbf{a}, \mathbf{b})$</li>
    <li>$d_{euklid}(\mathbf{a}, \mathbf{c})$</li>
    <li>$d_{manhattan}(\mathbf{a}, \mathbf{c})$</li>
    <li>$s_{cosine}(\mathbf{a}, \mathbf{c})$</li>
    <li>$d_{euklid}(\mathbf{c}, \mathbf{d})$</li>
    <li>$d_{manhattan}(\mathbf{c}, \mathbf{d})$</li>
    <li>$s_{cosine}(\mathbf{c}, \mathbf{d})$</li>
    <li>$d_{euklid}(\mathbf{a}, \mathbf{o})$</li>
    <li>$d_{manhattan}(\mathbf{a}, \mathbf{o})$</li>
    <li>$d_{euklid}(\mathbf{c}, \mathbf{o})$</li>
    <li>$d_{manhattan}(\mathbf{c}, \mathbf{o})$</li>
    </ul>
</div>


Lösung:

---
## <span style="color:red">Student Answer</span>

*Double-click and add your answer between the lines*

---


Anmerkung: Wir haben den euklidschen Abstand über die Norm, und die Norm über das Skalarprodukt definiert, obwohl das Ergebnis ja im karthesischen Koordinatenraum relativ naheliegend ist. Aber die Konstruktion ist allgemein - und damit ist klar, dass *jeder* Vektorraum mit Skalarprodukt notwendigerweise auch eine Norm und eine Abstandsmetrik hat.

Doch genug mit der Theorie.

<div class="aufgabe">
    <h3>Vektoren als Python-Klasse</h3>
    Implementieren Sie eine Klasse "Vector", die die üblichen Vector-Operationen und -Methoden 
    unterstützt. Unten finden Sie eine Liste von Methoden und einen Rahmen mit Unit-Tests für
    Ihren Code.
</div>

- `__init__(self, vec)`: Das Argument kann Tupel oder Liste von Gleitkommazahlen sein.
- `__eq__(self, other)`: Zwei Vektoren sind gleich, wenn sie die gleiche Anzahl von Komponenten haben, und die jeweils korrespondierenden Komponenten gleich sind.
- `__str__(self)`: Wir wollen Vektoren in der Form `<1, 2, 3>`  darstellen.
- `__repr__(self)`: Hier haben wir keine besonderen Anforderungen, Vorschlag: `Vector([1 2 3])`
- `__bool__(self)`: Nur der Null-Vektor (alle Komponenten 0) ist "False", alle anderen sind "True".
- `__len__(self)`: Anzahl der Dimensionen bzw. Komponenten.
- `__neg__(self)`: Gibt den Inversen Vektor zu `self` zurück.
- `__add__(self, other)`: Addiert zwei Vektoren komponentenweise.
- `__sub__(self, other)`: Subtrahiert zwei Vektoren.
- `__mul__(self, other)`: Multipliziert einen Vektor mit einem Skalar (z.B. <1, 0, 2>*3 = <3, 0, 6>) oder zwei Vektoren zum *Skalarprodukt*.
- `__rmul__(self, other)`: Das brauchen wir für den Fall, dass ein Skalar links an einen Vektor multipliziert wird (Fall `4*<1, 2, 3>=<4, 8, 12>).
- `is_zero(self)`: Ist `self` der Null-Vektor?
- `norm(self)`: Was ist die Länge im Euklidschen Raum?
- `euclid_dist(self, other)`: Was ist der Euklidsche Abstand zweier Vektoren?
- `manhattan_dist(self, other)`: Was ist der Manhattan-Abstand zweier Vektoren?
- `cosine_similarity(self, other)`: Was ist die Ähnlichkeit der Richtung zweier Vektoren?


In [3]:
# Notes:
# print(0.0 == 0) -> True

# print(vector.__len__()) for Vector([1,2,3]) -> 3
# print(repr(vector)) für repr

# Variablen aus NumPy Array:
# variables = [val for i, val in enumerate(self.vec)]

# Simples printing:
# for value in self.vec:
# print(value)

# technical_str returns:
# [np.int64(-1), np.int64(2), np.int64(3)]


# YOUR CODE HERE
class Vector:
    def __init__(self, vec):
        self.vec = np.array(vec)

    def check_dims(self, other):
        if not isinstance(other, Vector):
            raise VectorError("No Vector")
        if len(self.vec) != len(other.vec):
            raise VectorError("Vectors aren't the same size!")

    def __str__(self):
        return f"<{', '.join(str(x) for x in self.vec)}>"

    def __repr__(self):
        return f"Vector({self.vec.tolist()})"

    def __eq__(self, other):
        if not isinstance(other, Vector):
            return False
        if len(self.vec) != len(other.vec):
            return False
        return np.array_equal(self.vec, other.vec)

    def __len__(self):
        return len(self.vec)

    def __bool__(self):
        return bool(np.any(self.vec != 0))

    def __neg__(self):
        return Vector(-self.vec)

    def __add__(self, other):
        self.check_dims(other)
        return Vector(self.vec + other.vec)

    def __sub__(self, other):
        self.check_dims(other)
        return Vector(self.vec - other.vec)

    def __mul__(self, other):
        if isinstance(other, Vector):
            self.check_dims(other)
            return float(np.dot(self.vec, other.vec))
        return Vector(self.vec * other)

    def __rmul__(self, other):
        return Vector(other * self.vec)

    def is_zero(self):
        return np.all(self.vec == 0)

    def norm(self):
        return np.linalg.norm(self.vec)

    def euclid_dist(self, other):
        self.check_dims(other)
        return np.linalg.norm(self.vec - other.vec)

    def manhattan_dist(self, other):
        self.check_dims(other)
        return np.sum(np.abs(self.vec - other.vec))

    def cosine_similarity(self, other):
        self.check_dims(other)
        if self.is_zero() or other.is_zero():
            raise VectorError("Cannot compute cosine similarity with zero vector")
        return np.dot(self.vec, other.vec) / (self.norm() * other.norm())

#!/usr/bin/env python

"""Einführung Vektoren

In der Physik sind Vektoren Objekte, die durch eine Länge und eine
Richtung beschrieben werden. In der Mathematik sind Vektoren Elemente
eines Vektorraums. Ein Vektorraum ist eine mathematische Struktur,
konkret eine abelsche Gruppe (von "Vektoren") über einem Körper K (von
"Skalaren"), bei der eine zusätzliche Verknüpfung (die
Skalarmultiplikation) das Verknüpfen von Vektoren und Skalaren
ermöglicht. Für diese Skalarmultiplikation werden zusätzlich
Distributivgesetze und Assoziativgesetz gefordert.

In dieser Vorlesung bewegen wir uns zwischen diesen beiden
Definitionen. Für uns sind Vektoren Elemente von $R^n$, also $n$-Tupel
von reelen Zahlen. Dabei ist $n$ fest - d.h. jeder gegebene Vektorraum
enthält nur Vektoren mit $n$ Elementen.  Die reelen Zahlen R bilden
den Skalarkörper, und sind die zugehörigen Skalare.

Einfache Spezialfälle sind R^2 - dann reden wir über Vektoren in der
Ebene, und R^3 - Vektoren im 3-dimensionalen Raum. Diese beiden Fälle
haben eine Vielzahl von Anwendungen in Technik und Physik.

Wenn wir ein karthesisches Koordinatensystem annehmen, dann
repräsentiert der Vektor <1, 2, 3> die Verschiebung eines Punktes um
eine Einheit in der X-Koordinate, zwei Einheiten in der Y-Koordinate,
und drei Einheiten in der Z-Koordinate.

Analog können Vektoren aber auch selbst als Punkte in Raum
interpretiert werden. Der Übergang zwischen beiden Sichtweisen ist
einfach - der Vektor entspricht dem Punkt, auf dem er den
Koordinatenursprung verschieben würde.

"""

import unittest
import numpy as np
import math


class VectorError(Exception):
    """
    Für eigene Fehler, die mit Vektoren auftreten können.
    """
    pass

class TestVectors(unittest.TestCase):
    """
    Unittests für den Vector-Datentyp.
    """
    def setUp(self):
        """
        Initialisiere Variablen für den Test.
        """
        self.a  = Vector([0, 0, 1])
        self.b  = Vector([0, 1, 0])
        self.c  = Vector([1, 0, 0])
        self.d  = Vector([1, 1, 0])
        self.e  = Vector([1, 2, 3])
        self.f  = Vector([3, 2, 1])
        self.fn = Vector([-3, -2, -1])
        self.g  = Vector([-1, -2, 3])
        self.n  = Vector([0, 0, 0])
        self.o  = Vector([1, 1, 1])
        self.p  = Vector([1, 2])
        self.q  = Vector([1, 2, -3, 4])

    def test_01_basics(self):
        """
        Testet Basiseigenschaften - Gleichheit,
        Ausgabe.
        """
        self.assertEqual(self.a, self.a)
        self.assertNotEqual(self.a, self.b)
        self.assertEqual(str(self.e), "<1, 2, 3>")
        

    def test_02_dims(self):
        """
        Testet Dimensionen und den Umgang mit
        Unterschieden.
        """
        self.assertNotEqual(self.a, self.p)
        self.assertNotEqual(self.p, self.q)
        self.assertNotEqual(self.q, self.p)
        self.assertEqual(len(self.a), 3)
        self.assertEqual(len(self.p), 2)
        self.assertEqual(len(self.q), 4)

    def test_03_bool(self):
        """
        Testet die Umwandlung zu bool.
        """
        self.assertTrue(self.a)
        self.assertTrue(self.b)
        self.assertTrue(self.p)
        self.assertTrue(self.q)
        self.assertFalse(self.n)

    def test_04_addition(self):
        """
        Testet Addition.
        """
        self.assertEqual(self.b+self.c, self.d)
        self.assertEqual(self.f+self.n, self.f)
        self.assertEqual(self.f+self.fn, self.n)
        self.assertEqual(self.a+self.b+self.c, self.o)

    def test_05_subtraction(self):
        """
        Testet Subtraktion.
        """
        self.assertEqual(self.d-self.c, self.b)
        self.assertEqual(self.a-self.a, self.n)
        self.assertEqual(self.f-self.n, self.f)
        self.assertEqual(self.o-self.d, self.a)
        self.assertEqual(self.a-self.b-self.c, self.a-self.d)

    def test_06_negation(self):
        """
        Testet das unäre Minus.
        """
        self.assertEqual(-self.f, self.fn)
        self.assertEqual(-self.fn, self.f)
        self.assertEqual(self.a+-self.a, self.n)
        self.assertEqual(self.f+-self.n, self.f)
        self.assertEqual(self.d+-self.c, self.b)

    def test_07_multiplikation(self):
        """
        Testet Skalarmultiplikation (nicht Skalarprodukt!)
        """
        self.assertEqual(self.a, 1*self.a)
        self.assertEqual(self.a+self.a, 2*self.a)
        self.assertEqual(self.f*2, 2*self.f)
        self.assertNotEqual(self.a*2, self.a)
        self.assertEqual(self.a*0, self.n)

    def test_08_is_zero(self):
        """
        Testet is_zero().
        """
        self.assertFalse(self.a.is_zero())
        self.assertTrue(self.n.is_zero())
        
    def test_09_norm(self):
        """
        Testet  norm() und Skalarprodukt.
        """
        self.assertEqual(self.a.norm(),1)
        self.assertEqual(self.b.norm(),1)
        self.assertEqual(self.c.norm(),1)
        self.assertEqual(self.a*self.b, 0)
        self.assertEqual(self.a*self.a, 1)
        self.assertAlmostEqual(self.f*self.f, self.f.norm()*self.f.norm())
        self.assertAlmostEqual(self.o*self.o, self.o.norm()*self.o.norm())

    def test_10_similarity_euclid(self):
        """
        Testet Ähnlichkeitsmaß und Abstandsmaße.
        """
        self.assertEqual(self.a.euclid_dist(self.a),0)
        self.assertAlmostEqual(self.a.euclid_dist(self.b),1.4142135623730951)
        self.assertAlmostEqual(self.f.euclid_dist(self.fn),2*self.f.norm())

    def test_11_similarity_manhattan(self):
        """
        Testet Ähnlichkeitsmaß und Abstandsmaße.
        """
        self.assertEqual(self.a.manhattan_dist(self.a),0)
        self.assertEqual(self.a.manhattan_dist(self.b),2)
        self.assertAlmostEqual(self.f.manhattan_dist(self.fn),12)

    def test_12_similarity_cosine(self):
        """
        Testet Ähnlichkeitsmaß und Abstandsmaße.
        """
        self.assertEqual(self.a.cosine_similarity(self.a),1)
        self.assertEqual(self.a.cosine_similarity(self.b),0)
        self.assertAlmostEqual(self.f.cosine_similarity(self.fn),-1)
    

    def addition_dim_mismatch(self):
        self.p+self.q

    def addition_type_mismatch(self):
        self.p+"Kein Vektor"

    def multiplikation_dim_mismatch(self):
        self.p*self.q

    def cosine_similarity_with_zero(self):
        self.n.cosine_similarity(self.a)
    
    def test_13_error_handling(self):
        """
        Testet die Fehlerbehandlung
        """
        self.assertRaises(VectorError, self.addition_dim_mismatch)
        self.assertRaises(VectorError, self.addition_type_mismatch)
        self.assertRaises(VectorError, self.multiplikation_dim_mismatch)
        self.assertRaises(VectorError, self.cosine_similarity_with_zero)


if __name__ == '__main__':
    #Durchführung der Tests
    loader = unittest.TestLoader()
    suite = unittest.TestSuite()

    #Hier können einzelne Tests auskommentiert werden
    suite.addTest(TestVectors("test_01_basics"))
    suite.addTest(TestVectors("test_02_dims"))
    suite.addTest(TestVectors("test_03_bool"))
    suite.addTest(TestVectors("test_04_addition"))
    suite.addTest(TestVectors("test_05_subtraction"))
    suite.addTest(TestVectors("test_06_negation"))
    suite.addTest(TestVectors("test_07_multiplikation"))
    suite.addTest(TestVectors("test_08_is_zero"))
    suite.addTest(TestVectors("test_09_norm"))
    suite.addTest(TestVectors("test_10_similarity_euclid"))
    suite.addTest(TestVectors("test_11_similarity_manhattan"))
    suite.addTest(TestVectors("test_12_similarity_cosine"))
    suite.addTest(TestVectors("test_13_error_handling"))


    runner = unittest.TextTestRunner()

    runner.run(suite)
    '''
    print("Hacky unorganised tests ;-)")
    a = Vector([1, 2, 3])
    e = Vector([-7, 2, 1])
    b = Vector([3, 2, 1])
    c = Vector([1, 1, 0])
    d = Vector([0, 0, 1])
    n = Vector([0, 0, 0])
    print(a)
    print(repr(a))
    print(a+b)
    print(a+-a)
    print(a-a)
    print(a-b)
    print(a*a)
    print(a*b)
    print(4*a)
    print(a*4)
    print(a*n)
    print("0*a:", 0*a)
    print(c*d)
    print(a*e)
    print(a.norm())
    print(b.norm())
    print(n.norm())
    print(a.euclid_dist(b))
    print(a.euclid_dist(n))
    print(a.euclid_dist(a))
    print(a.manhattan_dist(b))
    print(a.cosine_similarity(a))
    print(a.cosine_similarity(b))
    print(c.cosine_similarity(d))
    print(c.cosine_similarity(-c))
    print(b.cosine_similarity(-b))
    '''

.............
----------------------------------------------------------------------
Ran 13 tests in 0.011s

OK


<div class="aufgabe">
    <h3>Bonus: Maschinelles Lernen mit <em>k</em>NN</h3>
    Implementieren Sie einen <em>k</em>NN-Algorithmus zur Klassifikation von Sportlern. Sie können obige Liste
    erweitern (Sportlerdaten findet man z.B. auf Wikipedia), um an mehr Trainings- und Testdaten zu kommen (Tipp:
    Dabei kann man sich gut austauschen, wenn jeder sich z.B. eine Sportart vornimmt). Vergleichen Sie folgende Fälle:
    <ul>
        <li>k=1, 3, 5</li>
        <li>Rohdaten vs. normalisierte Daten (zur Normalisierung skalieren Sie alle Werte auf den Zahlenbereich [0,1]).</li>
    </ul>
    Was funktioniert besser?
</div>


# Footer

In [ ]:
#Ausführen, um den aktuellen Footer anzuzeigen
from IPython.display import HTML
HTML(filename='files/footer.html')